src/lib/Microsoft.Health.Events/EventHubProcessor/PartitionLockingProcessor.cs

```diff
@@ -0,0 +1,418 @@
+        private int _maximumProcessors = 8;
+
+        private int _maximumPartitions = 8;
+
+        // settings that should be passed in as environment variables
+        private Uri _paritionManagerUri = new Uri("<blob uri goes here>");
+
+        private Uri _storageCheckpointUri = new Uri("<blob uri goes here>");
+
+        private string _eventHubConnectionString = "<connection string>";
+
+        private string _eventHubName = "<event hub name>";
+
+        private string _eventHubConsumerGroup = "$Default";
+
+        private TokenCredential _tokenCredential;
+
+        public PartitionLockingProcessor()
+        {
+            _processorId = Guid.NewGuid().ToString(); // consider using the pod name when running in AKS
+            _ownedPartitions = new List<int>();
+            _tokenCredential = new DefaultAzureCredential();
+        }
+
+        private void CreateContainerClient()
+        {
+            _containerClient = new BlobContainerClient(_paritionManagerUri, _tokenCredential);
+            _containerClient.CreateIfNotExists();
+        }
+
+        private async Task RegisterProcessor()
+        {
+            // write process id to storage
+            Console.WriteLine($"Processor Id: {_processorId}");
+            BlobClient blobClient = _containerClient.GetBlobClient($"activepods/{_processorId}");
+
+            var activeUntil = new Dictionary<string, string>()
+            {
+                { "activeuntil", DateTime.UtcNow.AddMinutes(1).ToString() },
+            };
+
+            BlobUploadOptions options = new BlobUploadOptions() { Metadata = activeUntil };
+            await blobClient.UploadAsync(BinaryData.FromString(string.Empty), options);
+
+            Console.WriteLine($"Registered processor {_processorId}");
+        }
+
+        private async Task RegisterProcessorLoop()
+        {
+            while (true)
+            {
+                await RegisterProcessor();
+                await Task.Delay(15000);
+            }
+        }
+
+        private async Task AcquirePartitions(CancellationToken ct)
+        {
+            // get all the pods in active status
+            var activePods = ListBlobs(_containerClient, "activepods");
+            var activePodCount = await GetActiveProcessors(activePods, _containerClient);
+            _totalProcessorsRunning = activePodCount;
+            Console.WriteLine($"Active Pod Count: {activePodCount}");
+
+            if (activePodCount == 0)
+            {
+                throw new Exception("active pods is 0");
+            }
+
+            // divide the partitions among the pods
+            // todo: handle cases where there is a remainder
+            // for example 8 partitions and 6 pods, which would require 4 pods to have 1 partition, and 2 pods to have 2 partitions
+            // for now, we will round down
+            var suggestedPartitionsPerPod = Math.Floor((double)_maximumProcessors / activePodCount);
```

Comment:
Conceptually this PR looks good. One thing to consider is that there will be cases when there are more active pods then `_maximumProcessors`. For example, during an upgrade K8s first creates new, upgraded pods and then terminates the old pods. So in that brief period of time there will not be enough partitions to go around. Not sure if this logic handles that case

Reply:
it is a good call out. I think the simplest approach is to still do the calculation to assign a suggested number of partitions per pod, and if it is zero for some amount of time due to an upgrade then that is fine. As long as we expect the number of pods to eventually reach the desired state (say 8), then when we periodically recalculate the number of suggested partitions per pod there should be enough to go around.

we may want to add an error condition for certain cases. like if there is one partition and we scale to 8 that could mean that no pods are assigned the one partition since we round down here. we probably shouldn't scale to 8 in that case either but the application should alert as well